app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt

```diff
@@ -0,0 +1,332 @@
+    }
+
+    private fun updateForwardButtonFocusTree(
+        focusedForwardButtonNode: FocusNode,
+        sessionState: SessionRepo.State
+    ): State {
+
+        assert(focusedForwardButtonNode.viewId == R.id.navButtonForward)
+
+        val nextFocusLeftId = when {
+            sessionState.backEnabled -> R.id.navButtonBack
+            else -> R.id.navButtonForward
+        }
+
+        return State(
+            focusNode = FocusNode(
+                focusedForwardButtonNode.viewId,
+                nextFocusLeftId = nextFocusLeftId))
+    }
+
+    private fun updatePocketMegaTileFocusTree(
+        focusedPocketMegatTileNode: FocusNode,
+        pinnedTilesIsEmpty: Boolean,
+        focused: Boolean = true
+    ): State {
+
+        assert(focusedPocketMegatTileNode.viewId == R.id.pocketVideoMegaTileView ||
+            focusedPocketMegatTileNode.viewId == R.id.megaTileTryAgainButton)
+
+        val nextFocusDownId = when {
+            pinnedTilesIsEmpty -> R.id.settingsTileContainer
+            else -> R.id.tileContainer
+        }
+
+        return State(
+            focusNode = FocusNode(
+                focusedPocketMegatTileNode.viewId,
+                nextFocusDownId = nextFocusDownId),
+            focused = focused)
+    }
+
+    /**
+     * Two possible scenarios for losing focus when in overlay:
+     * 1. When all the pinned tiles are removed, tilContainer no longer needs focus
+     * 2. When click on [megaTileTryAgainButton]
+     */
+    private fun handleLostFocusInOverlay(
+        lostFocusNode: FocusNode,
+        sessionState: SessionRepo.State,
+        pinnedTilesIsEmpty: Boolean,
+        pocketState: PocketVideoRepo.FeedState
+    ): State {
+
+        assert(lostFocusNode.viewId == R.id.tileContainer ||
+                lostFocusNode.viewId == R.id.megaTileTryAgainButton)
+
+        val viewId = when (pocketState) {
+            PocketVideoRepo.FeedState.FetchFailed -> R.id.megaTileTryAgainButton
+            PocketVideoRepo.FeedState.Inactive -> R.id.navUrlInput
+            else -> R.id.pocketVideoMegaTileView
+        }
+
+        val newFocusNode = FocusNode(viewId)
+
+        // If new focusNode is different from previous, set focused state to false
+        val focused = newFocusNode.viewId == lostFocusNode.viewId
+
+        return if (newFocusNode.viewId == R.id.navUrlInput) {
+            updateNavUrlInputFocusTree(newFocusNode, sessionState, pinnedTilesIsEmpty, pocketState, focused)
+        } else {
+            updatePocketMegaTileFocusTree(newFocusNode, pinnedTilesIsEmpty, focused)
+        }
+    }
+
+    private fun updateDefaultFocusForOverlayWhenTransitioningFromWebRender(
+        sessionState: SessionRepo.State
+    ) {
+        defaultFocusMap[ActiveScreen.NAVIGATION_OVERLAY] = when {
+            sessionState.backEnabled -> R.id.navButtonBack
+            sessionState.forwardEnabled -> R.id.navButtonForward
+            else -> R.id.navButtonReload
```

Comment:
Reload isn't guaranteed to always be enabled when returning from WebRender.  We need to fall back to navUrlInput if the current URL is home.

Reply:
`defaultFocusMap` is gone now, but yes good catch